### PR TITLE
Enable Alerts and stats events to UNIX-socket

### PR DESCRIPTION
**Pull request by psychogun**
Report triggered IDS/IPS alerts, internal performance counters of the Suricata IDS/IPS engine, such as captured traffic volume, memory usage, uptime, flow counters, and much more to a UNIX-socket.

This is to be used in conjunction with [[inputs.suricata]] which is a plugin for Suricata metrics in Telegraf.

### CODE REVIEW

**Comment by psychogun** [comment]
https://github.com/opnsense/plugins/pull/2494

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml [hunk 1]
```diff
@@ -70,6 +70,29 @@ stats:
...
 # Configure the type of alert (and other) logging you would like.
 outputs:
+{% if helpers.exists('OPNsense.IDS.general.unix_sock') and OPNsense.IDS.general.unix_sock == '1' %}
+
+  # Extensible Event Format (nicknamed EVE) to UNIX-socket
+  - eve-log:
+      enabled: yes
+      filetype: unix_stream
+      filename: /tmp/suricata-stats.sock
+      types:
```
Comment (AdSchellevis) [inline src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:80]: probably best to align output to the current syslog/local versions, e.g. implement payload options when `LogPayload` is enabled.

**Comment by AdSchellevis** [comment]
@psychogun some small remarks, but in general this looks like a nice addition.

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml [hunk 2]
```diff
@@ -70,6 +70,29 @@ stats:
 
 # Configure the type of alert (and other) logging you would like.
 outputs:
+{% if helpers.exists('OPNsense.IDS.general.unix_sock') and OPNsense.IDS.general.unix_sock == '1' %}
+
+  # Extensible Event Format (nicknamed EVE) to UNIX-socket
+  - eve-log:
+      enabled: yes
+      filetype: unix_stream
+      filename: /tmp/suricata-stats.sock
```
[the remark this answers is not in the document]
Comment (psychogun) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: Well, the problem is - Suricata does not create a UNIX-socket. It connects in on an already created UNIX-socket. This socket has to be created somewhere first, before this output starts working.

Telegraf is configured to create a listening socket (check the other pull request I linked to). Telegraf is started by the user `telegraf` and `telegraf` does not have permissions to `/var/log/suricata/suricata-eve-stats.sock`. 

I am open for solutions.

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml (hunk 1 above)
Comment (psychogun) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:80]: I am sorry, this I did not comprehend.

**Comment by psychogun** [comment]
Thank you for commenting my first ever pull request :)

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdSchellevis) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: hmm, you probably have to make sure the socket is created from the suricata startup as well in some way otherwise this would likely induce a race condition. I don't have a lot of time at the moment to look into this, but glueing services together is something we usually try to avoid (question in this case probably is how to make them loosly coupled)

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml (hunk 1 above)
Comment (AdSchellevis) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:80]: eve output is already defined twice (syslog and standard file), best make sure that the options about the log format are shared among all of them, in which case the log payload setting is missing in this section (just search for LogPayload in  the template and you see what I mean).

Comment (psychogun) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:80]: I think I understood. I have updated the section to include payload options.

**Comment by psychogun** [comment]
TL;DR
This feature can wait until Telegraf in OPNsense is updated with the ability to use [[inputs.suricata]] to produce metrics from alarms. 

And another remark, this output to a UNIX socket is desirable from my point of view, because the default processing using Telegraf to tail the `/var/log/suricata/eve.json` file does not work.  Telegraf (`telegraf`) does not have the correct permissions to access this file, thus no metrics are produced from Telegraf:  https://github.com/opnsense/plugins/issues/2239 

I have yet to receive a feedback from anyone regards to this matter, not on the forum nor on Github. 

And to further elaborate; stats and such from Suricata do produce metrics with the current version of Telegraf shipped with OPNsense, however, I am waiting of a new  version of Telegraf to test the ability to use [[inputs.suricata]] with the option to also produce metrics from **alerts** - https://github.com/influxdata/telegraf/tree/master/plugins/inputs/suricata

My main goal is to have the Alerts logged.

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdSchellevis) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: if startup order matters (which I didn't try), it would be a race, but if suricata starts normally and just keeps trying to connect, it's not a huge issue. I do however still prefer to keep suricata's sockets out of /tmp in case some would like to guard the files at some point in time.

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml (hunk 1 above)
Comment (AdSchellevis) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:80]: thanks, looks good

File: src/opnsense/service/templates/OPNsense/IDS/suricata.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (psychogun) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: It looks like Suricata just tries to reconnect the socket. 

Yes, I agree. World readable is probably not the best solution. However, as mentioned, this is a socket created by `telegraf` and the service Telegraf which Suricata connects to. 

The default suggestion is `source = "/var/run/suricata-stats.sock"`, but `telegraf` does not have permission to this folder. Or not that many other directories, either. 

So maybe `/var/log/telegraf/suricata-stats.sock` ..? I am very open for suggestions :)

[the remark this answers is not in the document]
Comment (AdSchellevis) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: I think /var/run is world readable/executable by default, you just have to make sure the created socket receives proper rights when created (which would be a concern for telegraf in your case).

[the remark this answers is not in the document]
Comment (psychogun) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: Ah, yes - `telegraf` is able to read / execute. But no bueno creating/writing a listening socket in this directory; 
```
root@opnsense:/var/log/telegraf # sudo -u telegraf nc -U -l /var/run/test.sock
nc: Permission denied
```

And I do not know how I would go about changing the default permissions of this socket when created by Telegraf either (if that is what you meant with the latest comment, to make it octal notation ~`750` (?));
```
root@opnsense:/var/log/telegraf # ls -alh /tmp/suricata-stats.sock 
srwxr-xr-x  1 telegraf  wheel     0B Aug 16 13:29 /tmp/suricata-stats.sock
```

I do not know exactly how UNIX sockets works, but; stopping Telegraf and opening a shell with `sudo -u telegraf nc -U -l /tmp/suricata-stats.sock` showed streams of data from Suricata in this shell. Opening another shell, where I tried to "hop" on in to this socket, as root, with `nc -l /tmp/suricata-stats.sock` to watch the same data stream (?), the output was nothing (I do not know if this is the proper way to "tap" in to an UNIX-socket though, in regards of leaving permissions of the created socket as-is, readable/executable for others). 

The socket `/tmp/suricata-stats.sock` does not increase in size or anything either while the stream is flowing.

`/tmp` uses sticky bit, that is why the group is `wheel`.

[the remark this answers is not in the document]
Comment (AdSchellevis) [reply src/opnsense/service/templates/OPNsense/IDS/suricata.yaml:79]: hmm, I think for the time being its best to keep these settings locally (in `/usr/local/etc/suricata/custom.yaml`), I currently don't have enough time to dig into how this should work and in order to add it as feature it's purpose should be consistent and clear.  The custom.yaml usually is a practical solution for more tailored setups.

**Comment by aimdev** [comment]
I agree with psychogun that this feature should be  provided.
I raised a comment in the forum regarding the issues re suricata and telegraf, see
https://forum.opnsense.org/index.php?topic=25352.msg121706#msg121706

**Comment by AdSchellevis** [comment]
The current problem is that `telegraf` doesn't seem to support enough ownership control to detach both subsystems reliably (https://github.com/influxdata/telegraf/issues/3239), an alternative might be to look into an udp transport with an interface selection from telegraf's end  https://www.influxdata.com/blog/telegraf-socket-listener-input-plugin/. 

In this case one could specify a valid (loopback) address in Suricata (or maybe even just pass it via the existing syslog subsystem) and have the service listen for messages.

**Comment by aimdev** [comment]
I am not sure I follow your argument, are you suggesting it is an issue to open a socket within the LAN environment, despite being protected by opnsense's firewall and bsd's secutity?
If so then the only alternative is to remove these functions from opnsense, and sniff the LAN forwarding the data to yet another system to monitor the network?

**Comment by AdSchellevis** [comment]
Quite simple, we could support a file_stream when properly documented, saved in a logic location (`/var/run`) and not tight to a specific service not part of core, but `telegraf` seems to have issues configuring ownership properly (there's a race condition) https://github.com/opnsense/core/pull/5151#discussion_r689471201

An alternative could be to use an udp stream in stead of a file stream as this might either work out of the box (using the available syslog targets) or is easy to implement (either direct from Suricata or preferably using syslog as this is the generic facility for log shipping).

**Comment by aimdev** [comment]
Well re my comment regarding persistence, perhaps you should allow the user the option of using the existing methods, while ensuring security for the majority of your users? Wiping out settings on boot/reboot/upgrade is counter-productive.

**Comment by AdSchellevis** [comment]
not sure what you mean, but custom persistent overwrites in Suricata are already possible using the `custom.yaml` file as explained earlier in this PR (https://github.com/opnsense/core/pull/5151#discussion_r690084873)

**Comment by aimdev** [comment]
> not sure what you mean, but custom persistent overwrites in Suricata are already possible using the `custom.yaml` file as explained earlier in this PR ([#5151 (comment)](https://github.com/opnsense/core/pull/5151#discussion_r690084873))

Please read the attached log, which shows any changes made to  custom.yml are overwitten if suricata is restarted via the gui, or opnsense is rebooted. 

# more /usr/local/etc/suricata/custom.yaml
%YAML 1.1
---
# empty stub for custom modifications, add custom persistent config below

## file edited

root@opnsense:~ # more /usr/local/etc/suricata/custom.yaml
%YAML 1.1
---
# empty stub for custom modifications, add custom persistent config below
- eve-log:
    enabled: yes
    filetype: unix_stream
    filename: /tmp/suricata-stats.sock
    types:
      - stats:
         threads: yes

## suricata disabled and reenabled using gui

root@opnsense:~ # more /usr/local/etc/suricata/custom.yaml
%YAML 1.1
---
# empty stub for custom modifications, add custom persistent config below

## custom.yml modified again
root@opnsense:~ # more /usr/local/etc/suricata/custom.yaml
%YAML 1.1
---
# empty stub for custom modifications, add custom persistent config below
- eve-log:
    enabled: yes
    filetype: unix_stream
    filename: /tmp/suricata-stats.sock
    types:
      - stats:
         threads: yes

##  opnsense rebooted

ssh root@opnsense
Password:
Last login: Wed Nov 17 05:05:37 2021 from 192.168.1.203
----------------------------------------------
|      Hello, this is OPNsense 21.7          |         @@@@@@@@@@@@@@@
|                                            |        @@@@         @@@@
| Website:	https://opnsense.org/        |         @@@\\\   ///@@@
| Handbook:	https://docs.opnsense.org/   |       ))))))))   ((((((((
| Forums:	https://forum.opnsense.org/  |         @@@///   \\\@@@
| Code:		https://github.com/opnsense  |        @@@@         @@@@
| Twitter:	https://twitter.com/opnsense |         @@@@@@@@@@@@@@@
----------------------------------------------

*** opnsense.sec.lan: OPNsense 21.7.5 (amd64/OpenSSL) ***


  0) Logout                              7) Ping host
  1) Assign interfaces                   8) Shell
  2) Set interface IP address            9) pfTop
  3) Reset the root password            10) Firewall log
  4) Reset to factory defaults          11) Reload all services
  5) Power off system                   12) Update from console
  6) Reboot system                      13) Restore a backup

Enter an option: 8

root@opnsense:~ # more /usr/local/etc/suricata/custom.yaml
%YAML 1.1
---
# empty stub for custom modifications, add custom persistent config below
root@opnsense:~ #

**Comment by fichtner** [comment]
It's simple really: you need to edit `/usr/local/opnsense/service/templates/OPNsense/IDS/custom.yaml` instead.


Cheers,
Franco

**Comment by aimdev** [comment]
> It's simple really: you need to edit `/usr/local/opnsense/service/templates/OPNsense/IDS/custom.yaml` instead.
> 
> Cheers, Franco

Thank You

**Comment by fichtner** [comment]
`/usr/local/opnsense/service/templates/OPNsense/IDS/custom.yaml` will not be overwritten ever. It will only be created if it does not exist during an update to make sure the stub is in place. The templates create the active configuration files in the system, but these local system files are always rewritten for consistency from the templates. The templates also have the advantage of being written in Jinja so you can have access to helper functionality, include files etc.

**Comment by AdSchellevis** [comment]
possible with a `custom.yaml` file, we'll keep it at that for now.
